isoutils/isoutils.go

```diff
@@ -0,0 +1,517 @@
+
+	var err error
+	options := args.Args{
+		SwupdMirror:             model.SwupdMirror,
+		SwupdStateDir:           tmpPaths[clr_initrd] + "/var/lib/swupd/",
+		SwupdStateClean:         true,
+		SwupdFormat:             "staging",
+		SwupdSkipDiskSpaceCheck: true,
+	}
+	sw := swupd.New(tmpPaths[clr_initrd], options)
+
+	/* Should install the overridden CoreBundles above (eg. os-core only) */
+	if err := sw.Verify(version, model.SwupdMirror, true); err != nil {
+		prg.Failure()
+		return err
+	}
+
+	prg.Success()
+	return err
+}
+
+func mkInitrdInitScript(templatePath string) error {
+	msg := "Creating and installing init script to initrd"
+	prg := progress.NewLoop(msg)
+	log.Info(msg)
+
+	type Modules struct {
+		Modules []string
+	}
+	mods := Modules{}
+
+	//Modules to insmod during init, paths relative to the kernel folder
+	modules := []string{
+		"/kernel/fs/isofs/isofs.ko",
+		"/kernel/drivers/cdrom/cdrom.ko",
+		"/kernel/drivers/scsi/sr_mod.ko",
+		"/kernel/fs/overlayfs/overlay.ko",
+	}
+
+	/* Find kernel, then break the name into kernelVersion */
+	kernelGlob, err := filepath.Glob(tmpPaths[clr_rootfs] + "/lib/kernel/org.clearlinux.native.*")
+	if err != nil || len(kernelGlob) != 1 {
+		prg.Failure()
+		log.Error("Failed to determine kernel revision")
+		return err
+	}
+	kernelVersion := strings.SplitAfter(filepath.Base((kernelGlob[0])), "org.clearlinux.native.")[1]
+
+	/* Copy files to initrd, and add to mods so they're added to the init template */
+	for _, i := range modules {
+		rootfsModPath := tmpPaths[clr_rootfs] + "/usr/lib/modules/" + kernelVersion + ".native" + i
+
+		/* copy kernel module to initramfs */
+		initrdModPath := filepath.Dir(tmpPaths[clr_initrd] + "/usr/lib/modules/" + kernelVersion + ".native" + i)
+		if _, err := os.Stat(initrdModPath); os.IsNotExist(err) {
+			os.MkdirAll(initrdModPath, os.ModePerm)
+		}
+
+		args := []string{"cp", "-rf", rootfsModPath, initrdModPath}
+		//out, err := exec.Command("sudo", args...).CombinedOutput()
+		err = cmd.RunAndLog(args...)
+		if err != nil {
+			prg.Failure()
+			log.Error("Failed to copy kernel modules")
+			return err
+		}
+
+		mods.Modules = append(mods.Modules, "/usr/lib/modules/"+kernelVersion+".native"+i)
+	}
+
+	tmpl, err := ioutil.ReadFile(templatePath + "/initrd_init_template")
+	if err != nil {
+		prg.Failure()
+		log.Error("Failed to open: initrd template at %s\n", templatePath+"initrd_init_template")
+		return err
+	}
+
+	t := template.New("Modules template")
+	t, err = t.Parse(string(tmpl))
+	if err != nil {
+		prg.Failure()
+		log.Error("Failed to parse init's template")
+		return err
+	}
+
+	f, err := os.Create(tmpPaths[clr_initrd] + "/init")
+	if err != nil {
+		prg.Failure()
+		log.Error("Failed to create init file for initrd!")
+		return err
+	}
+	defer f.Close()
+
+	err = t.Execute(f, mods)
+	if err != nil {
+		prg.Failure()
+		log.Error("Failed to execute template filling")
+		return err
+	}
+
+	/* Set correct owner and permissions on initrd's init */
```

Comment:
We should make these native go calls with os.Chmod and os.Chown

Reply:
Done